requests-handler.ts

```diff
@@ -0,0 +1,133 @@
+        logger.warn(`Request from: ${ctx.request.ip} failed because of DB error`);
+    }
+};
+
+const createProductRequest = async (ctx: ExtendableContext & DefaultContext & IRouterParamContext) => {
+    logger.info(`Got POST - create product request from ${ctx.request.ip}`);
+    try {
+        const value = await creationSchema.validateAsync(ctx.request.body);
+        ctx.status = 201;
+        const res = await createProduct(value);
+        ctx.body = res;
+        // @ts-ignore
+        logger.info(`New product created on DB by: ${ctx.request.ip}. Product ID: ${res._id}`)
+
+    } catch (err) {
+        if (err.name === 'MongoError') {
+            ctx.status = 500;
+            ctx.body = 'Internal Server Error';
+            logger.warn(`Request from: ${ctx.request.ip} failed because of DB error`);
+        } else {
+            ctx.status = 400;
+            ctx.body = err;
+        }
+    }
+};
+
+const updateProductRequest = async (ctx: ExtendableContext & DefaultContext & IRouterParamContext) => {
+    logger.info(`Got PUT - update product request from ${ctx.request.ip}`);
+    try {
+        const value = await updateSchema.validateAsync(ctx.request.body);
+        const res = await updateProduct(ctx.params.id, value);
+        if (res == null) {
+            ctx.body = 'Product Not Found!';
+            ctx.status = 404;
+        } else {
+            ctx.body = res;
+            // @ts-ignore
+            logger.info(`Product ID: ${res._id} updated by: ${ctx.request.ip}`)
+        }
+    } catch (err) {
+        if (err.name === 'MongoError') {
+            ctx.status = 500;
+            ctx.body = 'Internal Server Error';
+            logger.warn(`Request from: ${ctx.request.ip} failed because of DB error`);
+        } else {
+            ctx.status = 400;
+            ctx.body = err;
+        }
+    }
+};
+
+const deleteProductRequest = async (ctx: ExtendableContext & DefaultContext & IRouterParamContext) => {
+    logger.info(`Got DELETE - delete product request from ${ctx.request.ip}`);
+    try {
+        const res = await deleteProduct(ctx.params.id);
+        if (res == null) {
+            ctx.body = 'Product Not Found!';
+            ctx.status = 404;
+        } else {
+            ctx.body = res;
+            // @ts-ignore
+            logger.info(`Product ID: ${res._id} deleted by: ${ctx.request.ip}`)
+        }
+    } catch (err) {
+        ctx.status = 500;
+        ctx.body = 'Internal Server Error!';
+        logger.warn(`Request from: ${ctx.request.ip} failed because of DB error`);
+    }
+};
+
+const checkoutRequest = async (ctx: ExtendableContext & DefaultContext & IRouterParamContext) => {
+    logger.info(`Got POST - checkout request from ${ctx.request.ip}`);
+    try {
+        const value = await checkoutSchema.validateAsync(ctx.request.body);
+        const res = await checkout(value.buyList);
+        ctx.body = {success: res};
+    } catch (err) {
+        if (err.name === 'MongoError') {
```

Comment:
all your error handling code is the same.... find a way to write it only once

Reply:
I added error handler middleware. now the controllers only throws error upwards.
Please look if it's good enough